File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {  
    public boolean isSametree(TreeNode root1, TreeNode root2)
    {
        if(root1==null && root2==null) return true;
        if(root1!=null && root2==null || root1==null && root2!=null) return false;      
        return (root1.val==root2.val) && isSametree(root1.left,root2.left) && isSametree(root1.right,root2.right);
    }
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if(root==null) return false;
        boolean flag=false;
        if(isSametree(root,subRoot)) return true;
        flag=flag || isSubtree(root.left,subRoot);
        flag=flag || isSubtree(root.right,subRoot);    
        return flag;
     }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxHeight(self,root):

        if root == None:
            return 0
        res = 0
        stack = [[root,1]]
        while stack:
            node,depth = stack.pop()

            if node:
                res = max(depth,res)
                stack.append([node.left,depth+1])
                stack.append([node.right,depth + 1])
        return res
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if root == None:
            return True
        isBalanced = True
        stack = [root]

        while stack:
            node = stack.pop()
            if node:
                if abs(self.maxHeight(node.left)-self.maxHeight(node.right))>1:
                    isBalanced = False
                    break
                stack.append(node.left)
                stack.append(node.right)

        return isBalanced


        